### ISSUE TITLE
Absent `_mm_aesdec_si128` and `_mm_aesdeclast_si128`

### ISSUE THREAD

**Issue by norwend** [DLTcollab/sse2neon#477]
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_aesdec_si128&expand=262
https://developer.arm.com/architectures/instruction-sets/intrinsics/vaesdq_u8

**Comment by jserv** [comment]
Check [aes-brute-force/src/aes_ni_botan.cpp](https://github.com/GeneralZero/aes-brute-force/blob/dev/src/aes_ni_botan.cpp) at first glance.

- [ ] AES extension
```c
#define AES_DEC_4_ROUNDS(K) \
    do { \
        B0 = _mm_aesdec_si128(B0, K); \
        B1 = _mm_aesdec_si128(B1, K); \
        B2 = _mm_aesdec_si128(B2, K); \
        B3 = _mm_aesdec_si128(B3, K); \
    } while(0)
```

- [ ] Armv8-A Crypto extension
```c
#define AES_DEC_4_ROUNDS(K) \
    do { \
        B0 = vaesimcq_u8(vaesdq_u8(B0, K)); \
        B1 = vaesimcq_u8(vaesdq_u8(B1, K)); \
        B2 = vaesimcq_u8(vaesdq_u8(B2, K)); \
       B3 = vaesimcq_u8(vaesdq_u8(B3, K)); \
    } while(0)
```

**Comment by jserv** [comment]
Another AES-Crypto mapping example: https://gist.github.com/mmozeiko/f9c999dda7dbb03722409854a1c39cc2

**Comment by jserv** [comment]
IIRC, @wangxiao1254 implemented `_mm_aesimc_si128`, `_mm_aesdec_si128`, and `_mm_aesdeclast_si128` with ARMv8 Cryptography Extensions. See https://github.com/f1ed/emp/blob/sm4/emp-tool/utils/block.h

However, for SSE2NEON, we need ARMv7/non-crypto-ext counterparts.

**Comment by jserv** [comment]
Drop-in implementations with ARMv8 Cryptography Extensions:
```c
__m128i _mm_aesdec_si128 (__m128i a, __m128i RoundKey) {
    return vaesimcq_u8(vaesdq_u8(a, (__m128i){})) ^ RoundKey;
}
__m128i _mm_aesdeclast_si128 (__m128i a, __m128i RoundKey) {
    return vaesdq_u8(a, (__m128i){}) ^ RoundKey;
}
__m128i _mm_aesimc_si128 (__m128i a) {
    return vaesimcq_u8(a);
}
```

**Comment by jserv** [comment]
> However, for SSE2NEON, we need ARMv7/non-crypto-ext counterparts.

The portable implementations:
* QEMU [target/i386/ops_sse.h](https://github.com/qemu/qemu/blob/master/target/i386/ops_sse.h): Check `helper_aesdec` and `helper_aesdeclast`.
* [WebAssembly/design Issue #1433](https://github.com/WebAssembly/design/issues/1433)

**Comment by jserv** [comment]
`_mm_aesdec_si128` is implemented in commit 0f28c2539e950cd8dde2f042c39b87cd61c82937 .
